File: path-sum/path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        targetAchieved = False
        
        def dfs(node, sum):
            nonlocal targetAchieved
            
            if targetAchieved:
                return
            
            if not node:
                return
            
            sum += node.val
            
            if not node.left and not node.right:
                if sum == targetSum:
                    targetAchieved = True
                    return
                
            if node.left:
                dfs(node.left, sum)
            if node.right:
                dfs(node.right, sum)
                
                
        dfs(root, 0)
        
        return targetAchieved

File: path-sum/path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        return dfs(root, targetSum, 0);
    }
    
    public boolean dfs(TreeNode node, int targetSum, int sum) {
        boolean found = false;
        
        if (node != null) {
            
            sum += node.val;
            
            if (node.left == null && node.right == null) {
                if(sum == targetSum) {
                    found = true;
                }
            }
            
            if (node.left != null) {
                found = found || dfs(node.left, targetSum, sum);
            }
            
            if (node.right != null) {
                found = found || dfs(node.right, targetSum, sum);
            }
        }
        
        return found;
    }
}